whatshap/cli/extend.py

```diff
@@ -0,0 +1,238 @@
+                homozygous_number = 0
+                phased_number = 0
+                for variant, (phase, genotype) in zip(
+                    variant_table.variants, zip(phases, genotypes)
+                ):
+                    homozygous[variant.position] = genotype.is_homozygous()
+                    phased[variant.position] = phase
+                    phased_number += phase is not None
+                    homozygous_number += genotype.is_homozygous()
+                    change[variant.position] = variant
+                logger.info(f"Number of homozygous variants is {homozygous_number}")
+                logger.info(f"Number of already phased variants is {phased_number}")
+                with timers("compute-votes"):
+                    votes = compute_votes(homozygous, reads)
+                with timers("compute-consensus"):
+                    sample_to_super_reads[sample], sample_to_components[sample] = consensus(
+                        only_indels, gap_threshold, cut_poly, fasta_chr, change, phased, votes
+                    )
+            with timers("write-vcf"):
+                vcf_writer.write(chromosome, sample_to_super_reads, sample_to_components)
+    timers.stop("extend-run")
+    log_time_and_memory_usage(timers)
+
+
+def log_time_and_memory_usage(timers):
+    logger.info("\n# Resource usage")
+    log_memory_usage()
+    # fmt: off
+    logger.info("Finished in :                              %6.1f s", timers.elapsed("extend-run"))
+    logger.info("Time spent reading reference:              %6.1f s", timers.elapsed("read-fasta"))
+    logger.info("Time spent reading VCF:                    %6.1f s", timers.elapsed("parse-vcf"))
+    logger.info("Time spent writing VCF:                    %6.1f s", timers.elapsed("write-vcf"))
+    logger.info("Time spent reading BAM:                    %6.1f s", timers.elapsed("read-bam"))
+    logger.info("Time spent computing votes:                %6.1f s", timers.elapsed("compute-votes"))
+    logger.info("Time spent spent computing consensus:      %6.1f s", timers.elapsed("compute-consensus"))
+    # fmt: on
+
+
+def consensus(only_indels, gap_threshold, cut_poly, fasta_chr, change, phased, votes):
+    super_reads = [[], []]
+    components = dict()
+
+    for pos, var in votes.items():
+        al1, q, score1 = best_candidate(components, pos, var)
+        if phased[pos] is None:
+            if 100 * q < gap_threshold:
+                continue
+            if only_indels and change[pos].is_snv():
+                continue
+            if cut_poly > 0:
+                max_length = max(
+                    length_of_polymer(fasta_chr, pos + 1, 1, cut_poly),
+                    length_of_polymer(fasta_chr, pos, -1, cut_poly),
+                )
+                if max_length >= cut_poly:
+                    continue
+        super_reads[0].append(Variant(pos, allele=al1, quality=score1))
+        super_reads[1].append(Variant(pos, allele=al1 ^ 1, quality=score1))
```

Comment:
XORing with 1 is the same as NOT. Is the intention to map 1 to 0 and 0 to 1? Then please use `1 - al1`. Or `int(not al1)` maybe.